### PR TITLE
Preserves shape when 0 dim tensors are replicated

### LINKED ISSUE #167524: `nn.DataParallel` models failed to copy parameters correctly when 0 involved in shape
### 🐛 Describe the bug

When there's a `0` in `nn.Parameter`'s shape (e.g. 1x0x16, 0x4x32. Though it seems weird, such parameters are convenient when handling optional params with unfixed shape), the shape of `Parameter` will incorrectly change when copied to different devices. 

A minimal example: 
```python
import torch as pt

class ExampleModule(pt.nn.Module): 
    def __init__(self, num_extra_modules: int = 3):
        pt.nn.Module.__init__(self)
        self.base_vec = pt.nn.Parameter(pt.zeros(1, 16))
        self.extra_vec = pt.nn.Parameter(pt.zeros(num_extra_modules, 16))
    
    def forward(self, x: pt.Tensor): 
        print(f"Shape of `self.extra_vec`: {self.extra_vec.shape} on {self.extra_vec.device}")
        return x + pt.concat([self.base_vec, self.extra_vec], dim = 0).sum(0)

model_cpu = ExampleModule(0)
model_parallel = pt.nn.DataParallel(model_cpu, [0, 1])
model_parallel.to(pt.device("cuda:0"))
model_parallel(pt.ones(4, 16).to("cuda:0"))
```

Expected output: 
```
Shape of `self.extra_vec`: torch.Size([0, 16]) on cuda:0
Shape of `self.extra_vec`: torch.Size([0, 16]) on cuda:1
```

Actual output: 
```
Shape of `self.extra_vec`: torch.Size([0, 16]) on cuda:0
Shape of `self.extra_vec`: torch.Size([0]) on cuda:1
```

### Versions

Collecting environment information...
PyTorch version: 2.1.1
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 专业版 # NOTE: 专业版-> Professional
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.27.4
Libc version: N/A

Python version: 3.9.18 (main, Sep 11 2023, 14:09:26) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19045-SP0
Is CUDA available: True
CUDA runtime version: 12.4.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 4090
GPU 1: NVIDIA GeForce RTX 4090

Nvidia driver version: 560.94
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture=9
CurrentClockSpeed=2700
DeviceID=CPU0
Family=179
L2CacheSize=48640
L2CacheSpeed=
Manufacturer=GenuineIntel
MaxClockSpeed=2700
Name=Intel(R) Xeon(R) Platinum 8374B CPU @ 2.70GHz
ProcessorType=3
Revision=27142

Architecture=9
CurrentClockSpeed=2700
DeviceID=CPU1
Family=179
L2CacheSize=48640
L2CacheSpeed=
Manufacturer=GenuineIntel
MaxClockSpeed=2700
Name=Intel(R) Xeon(R) Platinum 8374B CPU @ 2.70GHz
ProcessorType=3
Revision=27142

Versions of relevant libraries:
[pip3] mypy==1.9.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.0
[pip3] torch==2.1.1
[pip3] torchaudio==2.1.1
[pip3] torchstat==0.0.7
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.16.1
[conda] blas                      1.0                         mkl    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl                       2023.1.0         h6b88ed4_46357    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl-service               2.4.0            py39h2bbff1b_1    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl_fft                   1.3.8            py39h2bbff1b_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl_random                1.2.4            py39h59b6b97_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] numpy                     1.26.0           py39h055cbcc_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] numpy-base                1.26.0           py39h65a83cf_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] pytorch                   2.1.1           py3.9_cuda12.1_cudnn8_0    pytorch
[conda] pytorch-cuda              12.1                 hde6ce7c_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.1.1                    pypi_0    pypi
[conda] torchstat                 0.0.7                    pypi_0    pypi
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.16.1                   pypi_0    pypi

### PR DESCRIPTION
**Pull request by arkadip-maitra**
Fixes #167524

### CODE REVIEW

**Comment by wconstab** [comment]
Seems ok to me. Thanks for the fix!

**Review by wconstab — APPROVED** [review]
